content/2019-02-19_osmnacty-sraz-git.md

```diff
@@ -0,0 +1,71 @@
+Více k tomuto tématu viz [dokumentace](https://git-scm.com/docs/git-cherry-pick).
+
+### git reset
+
+Příkaz `git reset` umí rušit commity. Pokud za příkaz zadáme název commitu, přesuneme se na něj a tedy v rámci gitu do stavu, v jakém tehdy repozitář byl. Změny, které se staly v pozdějších commitech, nejsou nyní v gitu, ale existují v našem repozitáři. Pokud se jich chci zbavit a úplně se vrátit do předchozího stavu, existují dvě cesty. Buď po `git reset "název commitu"` zadám ještě `git checkout .` Druhá cesta je přímější, a to pomocí `git reset --hard "název commitu"` přímo pozmění repozitář do stavu v daném commitu a vše, co se stalo po něm, zahodí, včetně rozpracovaných věcí. 
+Pomocníkem při pohybech po commitech může být `git reflog`. Git si pamatuje, jak se posouval ukazatel `HEAD` v rámci repozitáře, a `git reflog` vypíše historii všech těchto pohybů, včetně vymazaných commitů a také včetně provedených příkazů, které k pohybu vedly. Historie se ale čas od času promazává, nelze tedy jít libovolně dávno do minulosti.
+
+![git_reflog](./images/reflog.png)
+
+V rámci gitu se pohybjeme v podstatě ve třech úrovních - viz obrázek níže. Vycházíme z našeho pracovního adresáře, kde máme vztvořený gitový repozitář. Dokud gitu neřekneme pomocí `git add`, aby sledoval nějaké změny, červenou barvou nám po zadání `git status` vypíše soubory, které v adresáři jsou, ale nesleduje je. Když vložíme soubor do gitu, přesune se do zelené oblasti, tzv. Stage, kde je již sleduje, ale stále s nimi můžeme pracovat, měnit je apod. 
+Pokud máme soubor ve Stagi, ale chceme jej odebrat a vrátit zpět do pracovního adresáře, lze to pomocí `git reset -- "název souboru"`. 
+Ze Stage pak pomocí příkazu `git commit` git sledované soubory zabalí, dá jim identifikátor, hlavičku, datum, email apod., vytvoří commit a uloží ho. 
+
+![git_colors](./images/git_colors.png)
+*[Zdroj: Začátečnický kurz PyLadies](https://pyvec.github.io/cheatsheets/basic-git/basic-git-cs.pdf)*
+
+### git rebase
+
+Příkaz `git rebase` vykonává opakovaný `git cherry-pick`. Zadáme-li `git rebase "název větve"`, git se podívá, jak se od sebe dané větve liší a jaké jsou mezi nimi rozdílné commity. Postupně pak jednotlivé commity vezme a provede `git cherry-pick`. Důležité je pořadí větví. Ta větev, na které aktuálně jsem, slouží jako zdroj commit. Větev, kterou předáme jako parametr příkazu `git rebase`, bude ta, na kterou budou commity přibývat. 
+Díky `git rebase` můžeme naši větěv přesouvat libovolně po "stromě" našeho projektu, tedy vlastně přepisovat jeho historii, narozdíl od `git merge`, který větve a změny v nich sloučí do jedné, ale jejich historii nechá paralelní. 
+Pomocí `git rebase -i` se dostaneme do interaktivního módu, kde je možné měnit, co se pomocí `git rebase` má stát. Lze úplně vymazat některé commity, změnit commit message, upravit commit před jeho začleněním apod. Interaktivní mód obsahuje seznam commitů spolu s tím, jak se mají začlenit, a nápovědu, jak provést konkrétní úpravy.
```

Comment:
tady by bylo dobré ještě zmínit, že pomoci interaktivního rebase měnit i pořadí commitů nebo je i zcela smazat - klidně z prostředka historie

Reply:
Však to tam je :) Tedy to mazání. Přišlo mi zbytečné vypisovat všechny možnosti. Ale přidám tedy i změnu pořadí.

Reply:
ajo, vymazání jsem přehlédl - beru zpět.